### PR TITLE
:hammer: Rework for Probabilistic Regression & Pixel Regression

### PR DESCRIPTION
**Pull request by alafage**
- [x] Rework Distribution Layers to be classifiers returning distribution parameters in a dictionary.
- [x] Update RegressionRoutine
- [x] Update PixelRegressionRoutine
- [x] Update Tutorials
- [x] Update API Reference
- [x] Fix DistributionNLL (loss / metric)
- [x] Update BTS baseline
- [x] Add tutorial for simple probabilistic regression
- [x] Update existing tests
- [x] Write tests

### CODE REVIEW

File: torch_uncertainty/layers/distributions.py [hunk 1]
```diff
@@ -1,110 +1,506 @@
-from abc import ABC, abstractmethod
-
+import torch
 import torch.nn.functional as F
 from torch import Tensor, nn
-from torch.distributions import Distribution, Laplace, Normal
 
-from torch_uncertainty.utils.distributions import NormalInverseGamma
 
+def get_dist_linear_layer(dist_family: str) -> type[nn.Module]:
```
Comment (o-laurent) [inline torch_uncertainty/layers/distributions.py:6]: better than a class with __new__?

Comment (alafage) [reply torch_uncertainty/layers/distributions.py:6]: I think it's simpler than using new in a class only to return a type. But I don't mind changing it if you think it's better.

File: auto_tutorials_source/tutorial_probabilistic_regression.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
+"""
+Deep Probabilistic Regression
```
Comment (o-laurent) [inline auto_tutorials_source/tutorial_probabilistic_regression.py:2]: Seems great!

File: experiments/regression/uci_datasets/configs/boston/mlp/laplace.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+eval_after_fit: true
+trainer:
+  accelerator: gpu
+  devices: 1
+  precision: 16-mixed
+  max_epochs: 40
+  logger:
+    class_path: lightning.pytorch.loggers.TensorBoardLogger
+    init_args:
+      save_dir: logs/boston/mlp/laplace
```
Comment (o-laurent) [inline experiments/regression/uci_datasets/configs/boston/mlp/laplace.yaml:12]: We should avoid using this dataset though

File: experiments/regression/uci_datasets/readme.md [hunk 4]
```diff
@@ -1,17 +1,33 @@
...
+| Kin8nm | 8192 | 8 |
+| Naval Propulsion | 11,934 | 16 |
+| Combined Cycle Power Plant | 9568 | 4 |
+| Protein Structure | 45730 | 9 |
+| Wine Quality (Red) | 1599 | 11 |
+| Yacht Hydrodynamics | 308 | 6 |
+
+
+> [!WARNING]
+> Some datasets require installing additional packages.
```
Comment (o-laurent) [inline experiments/regression/uci_datasets/readme.md:18]: We could list them here, or tell to download optional dependencies with pip install torch_uncertainty[xxx]

File: torch_uncertainty/baselines/regression/mlp.py [hunk 5]
```diff
@@ -30,37 +25,19 @@ def __init__(
         dropout_rate: float = 0.0,
         alpha: float | None = None,
         gamma: int = 1,
-        distribution: Literal["normal", "laplace", "nig"] | None = None,
+        dist_family: str | None = None,
+        dist_args: dict | None = None,
     ) -> None:
         r"""MLP baseline for regression providing support for various versions."""
```
Comment (o-laurent) [inline torch_uncertainty/baselines/regression/mlp.py:31]: :book:

File: torch_uncertainty/losses/bayesian.py [hunk 6]
```diff
@@ -48,6 +51,8 @@ def __init__(
             inner_loss (nn.Module): The loss function to use during training
             kl_weight (float): The weight of the KL divergence term
             num_samples (int): The number of samples to use for the ELBO loss
+            dist_family (str, optional): The distribution family to use for the
+                output of the model. None means no distribution. Defaults to None.
```
Comment (o-laurent) [inline torch_uncertainty/losses/bayesian.py:55]: means pointwise prediction?

File: torch_uncertainty/models/depth/bts.py [hunk 7]
```diff
@@ -291,7 +290,7 @@ def __init__(
         max_depth: float,
         feat_out_channels: list[int],
         num_features: int,
-        dist_layer: type[nn.Module],
+        dist_family: str | None = None,
     ):
```
Comment (o-laurent) [inline torch_uncertainty/models/depth/bts.py:294]: :book:

File: torch_uncertainty/models/depth/bts.py [hunk 8]
```diff
@@ -486,10 +493,15 @@ def forward(self, features: list[Tensor]) -> Tensor | Distribution:
         """
```
Comment (o-laurent) [inline torch_uncertainty/models/depth/bts.py:493]: precise the output which depends on :attr: dist_family?

File: torch_uncertainty/models/mlp.py [hunk 9]
```diff
@@ -32,9 +33,9 @@ def __init__(
             layer (nn.Module): Layer class.
             activation (Callable): Activation function.
             layer_args (Dict): Arguments for the layer class.
-            final_layer (nn.Module): Final layer class for distribution regression.
-            final_layer_args (Dict): Arguments for the final layer class.
             dropout_rate (float): Dropout probability.
+            dist_family (str, optional): Distribution family. Defaults to None.
```
Comment (o-laurent) [inline torch_uncertainty/models/mlp.py:37]: detail what is None

File: torch_uncertainty/models/wrappers/deep_ensembles.py [hunk 10]
```diff
@@ -42,17 +39,18 @@ def __init__(
...
 
-    def forward(self, x: torch.Tensor) -> Distribution:
-        r"""Return the logits of the ensemble.
+    def forward(self, x: torch.Tensor) -> torch.Tensor | dict[str, torch.Tensor]:
+        """Return the logits of the ensemble.
 
         Args:
             x (Tensor): The input of the model.
-
-        Returns:
```
Comment (o-laurent) [inline torch_uncertainty/models/wrappers/deep_ensembles.py:51]: Detail the return type?

File: torch_uncertainty/routines/pixel_regression.py [hunk 11]
```diff
@@ -64,9 +62,9 @@ def __init__(
         Args:
             model (nn.Module): Model to train.
             output_dim (int): Number of outputs of the model.
-            probabilistic (bool): Whether the model is probabilistic, i.e.,
-                outputs a PyTorch distribution.
             loss (nn.Module): Loss function to optimize the :attr:`model`.
+            dist_family (str, optional): The distribution family to use for
+                probabilistic pixel regression. Defaults to ``None``.
```
Comment (o-laurent) [inline torch_uncertainty/routines/pixel_regression.py:67]: If None, pointwise regression

File: torch_uncertainty/routines/pixel_regression.py [hunk 12]
```diff
@@ -174,39 +173,47 @@ def training_step(self, batch: tuple[Tensor, Tensor], batch_idx: int) -> STEP_OU
...
-        out_shape = dist_size(dists)[-2:] if self.probabilistic else dists.shape[-2:]
+        out = self.model(inputs)
+        out_shape = out[next(iter(out))].shape[-2:] if self.probabilistic else out.shape[-2:]
         target = F.resize(target, out_shape, interpolation=F.InterpolationMode.NEAREST)
-        padding_mask = torch.isnan(target)
+        target = rearrange(target, "b c h w -> b h w c")
+        padding_mask = torch.isnan(target).any(dim=-1)
         if self.probabilistic:
+            dist_params = {k: rearrange(v, "b c h w -> b h w c") for k, v in out.items()}
+            dists = Independent(get_dist_class(self.dist_family)(**dist_params), 1)
```
Comment (o-laurent) [inline torch_uncertainty/routines/pixel_regression.py:183]: Add a comment to explain why we need Independent (we're not all experts on torch.distributions :smile: )

File: torch_uncertainty/routines/regression.py [hunk 13]
```diff
@@ -76,7 +79,9 @@ def __init__(
             )
 
         self.model = model
-        self.probabilistic = probabilistic
+        self.dist_family = dist_family
+        self.dist_estimate = dist_estimate
```
Comment (o-laurent) [inline torch_uncertainty/routines/regression.py:83]: Why is `self.dist_estimate` here and not in pixel-wise regression?

File: torch_uncertainty/layers/distributions.py (hunk 1 above)
Comment (o-laurent) [reply torch_uncertainty/layers/distributions.py:6]: Maybe it is, yes!

**Review by o-laurent — APPROVED** [review]
Some comments on the documentation mainly! 
Thanks for improving this part that wasn't that great (Sorry :zipper_mouth_face: )

File: torch_uncertainty/routines/regression.py (hunk 13 above)
Comment (alafage) [reply torch_uncertainty/routines/regression.py:83]: That's a mistake, thanks!

File: torch_uncertainty/losses/bayesian.py [hunk 14]
```diff
@@ -74,8 +80,10 @@ def forward(self, inputs: Tensor, targets: Tensor) -> Tensor:
         """
         aggregated_elbo = torch.zeros(1, device=inputs.device)
         for _ in range(self.num_samples):
-            logits = self.model(inputs)
-            aggregated_elbo += self.inner_loss(logits, targets)
+            out = self.model(inputs)
+            if self.dist_family is not None:
+                out = Independent(get_dist_class(self.dist_family)(**out), 1)
```
[the remark this answers is not in the document]
Comment (alafage) [reply torch_uncertainty/losses/bayesian.py:85]: Not at all, I'll fix it

[the remark this answers is not in the document]
Comment (o-laurent) [reply torch_uncertainty/losses/bayesian.py]: Maybe also recall why we need `Independent`?

**Comment by alafage** [comment]
Thank you for the great review! This last commit should address all your worries. Feel free to merge it into dev anytime!

File: torch_uncertainty/models/depth/bts.py [hunk 15]
```diff
@@ -488,8 +498,8 @@ def forward(self, features: list[Tensor]) -> Tensor | dict[str, Tensor]:
             features (list[Tensor]): list of the features from the backbone.
 
         Note:
-            Depending of the :attr:`dist_layer` of the backbone, the output can
-            be a distribution or a single tensor.
+            Depending of the :attr:`dist_family` of the backbone, the output can
+            be a dictionnary of distribution parameters or a single tensor.
```
Comment (o-laurent) [inline torch_uncertainty/models/depth/bts.py:502]: dictionary

**Review by o-laurent — APPROVED** [review]
Thanks for the changes! Just saw a small typo but we're good to go!
